src/Processors/QueryPlan/PartsSplitter.h

```diff
@@ -0,0 +1,322 @@
+                        event_queue.emplace(std::move(new_begin_value), part_idx, subrange_for_next_layer);
+                    }
+                }
+                started_ranges.clear();
+                result_layers.emplace_back();
+            }
+        }
+        for (const auto & range : started_ranges)
+            result_layers.back().emplace_back(parts[range.part_idx].data_part, range.part_idx, MarkRanges{range});
+
+        for (size_t layer = 0; layer < result_layers.size(); ++layer)
+        {
+            LOG_DEBUG(&Poco::Logger::get("PartsSplitter"), "Layer {}, ranges:", layer);
+            for (const auto & range : result_layers[layer])
+                for (const auto & mark_range : range.ranges)
+                    LOG_DEBUG(
+                        &Poco::Logger::get("PartsSplitter"),
+                        "part_idx={}, begin={}, end={}",
+                        range.part_index_in_query,
+                        mark_range.begin,
+                        mark_range.end);
+        }
+
+        return result_layers;
+    }
+
+    /// Will return borders.size()+1 filters in total, i-th filter will accept values from the range [borders[i-1], borders[i]).
+    std::vector<ASTPtr> buildFilters(const std::vector<IndexValue> & borders)
+    {
+        auto add_and_condition = [&](ASTPtr & result, const ASTPtr & foo) { result = !result ? foo : makeASTFunction("and", result, foo); };
+        auto add_or_condition = [&](ASTPtr & result, const ASTPtr & foo) { result = !result ? foo : makeASTFunction("or", result, foo); };
+
+        const auto & types = primary_key.data_types;
+        auto lexicographically_less = [&](const IndexValue & value)
+        {
+            ASTPtr result;
+            ASTPtr prefix_of_equal;
+            for (size_t i = 0; i < value.size(); ++i)
+            {
+                auto primary_key_column_ast = primary_key.expression_list_ast->children.at(i);
+                ASTPtr value_ast = std::make_shared<ASTLiteral>(value[i]);
+                if (isColumnedAsNumber(types.at(i)->getTypeId()) && !isNumber(types.at(i)->getTypeId()))
+                    value_ast = makeASTFunction("cast", std::move(value_ast), std::make_shared<ASTLiteral>(types.at(i)->getName()));
+
+                ASTPtr current = makeASTFunction("less", primary_key_column_ast, value_ast); // pk_col[i] < value[i]
+                if (prefix_of_equal)
+                    add_and_condition(
+                        current, prefix_of_equal); // pk_col[i] < value[i] && pk_col[i-1] == value[i-1] && ... && pk_col[0] == value[0]
+                add_or_condition(
+                    result,
+                    current); // (pk_col[i] < value[i] && pk_col[i-1] == value[i-1] && ... ) || same for pk_col[i-1] || ... || same for pk_col[0]
+
+                add_and_condition(prefix_of_equal, makeASTFunction("equals", primary_key_column_ast, value_ast));
+            }
```

Comment:
It looks inefficient (~N^2 comparisons for PK size = N?)
I think we can compare tuples here: `tuple(PK columns) <= tuple(value[0], value[1], ... value[N - 1])`.
But, still, I would prefer adding a special transform which can work directly with borders.